### ISSUE TITLE
Extending to batch sizes > 1

### ISSUE THREAD

**Issue by LukasBommes** [moganesyan/tensorflow_model_deployment#9]
Dear Mikheil,

thank you very much for providing this repository. It is very helpful and I got it working without any issues.

However, I want to extend it to batche sizes larger one. I looked into the batching functionality of tf serving, but  ran into issues. Thus, I decided to post a pre-assembled batch of images and let the served model perform the inference on the entire batch.

To this end, I modified the preprocess_input function as follows (basically copying the contents of the `detect` method of the Mask R-CNN model):
```
def preprocess_input(self, images):
        """Pre-processes the input image.
        img: Input image of shape (-1,XX,YY,3)
        Returns:
        molded_image: Molded image to be used as model input
        image_meta: Input image metadata
        anchors: [N, (y1, x1, y2, x2)]. All generated anchors in one array. Sorted
            with the same order of the given scales. So, anchors of scale[0] come
            first, then anchors of scale[1], and so on.
        window: (y1, x1, y2, x2). If max_dim is provided, padding might
            be inserted in the returned image. If so, this window is the
            coordinates of the image part of the full image (excluding
            the padding). The x2, y2 pixels are not included.
        """

        assert len(images) == self._config.BATCH_SIZE

        # Mold inputs to format expected by the neural network
        molded_images, image_metas, windows = self.mold_inputs(images)

        # Validate image sizes
        # All images in a batch MUST be of the same size
        image_shape = molded_images[0].shape
        for g in molded_images[1:]:
            assert g.shape == image_shape,\
                "After resizing, all images must have the same size. Check IMAGE_RESIZE_MODE and image sizes."

        # Anchors
        anchors = self.get_anchors(image_shape)
        # Duplicate across the batch dimension because Keras requires it
        # TODO: can this be optimized to avoid duplicating the anchors?
        anchors = np.broadcast_to(anchors, (self._config.BATCH_SIZE,) + anchors.shape)

        return molded_images, image_metas, anchors, windows
```
With `mold_inputs` as follows:

```
def mold_inputs(self, images):
        """Takes a list of images and modifies them to the format expected
        as an input to the neural network.
        images: List of image matrices [height,width,depth]. Images can have
            different sizes.

        Returns 3 Numpy matrices:
        molded_images: [N, h, w, 3]. Images resized and normalized.
        image_metas: [N, length of meta data]. Details about each image.
        windows: [N, (y1, x1, y2, x2)]. The portion of the image that has the
            original image (padding excluded).
        """
        molded_images = []
        image_metas = []
        windows = []
        for image in images:
            molded_image, window, scale, padding, crop = self.resize_image(
                image,
                min_dim=self._config.IMAGE_MIN_DIM,
                min_scale=self._config.IMAGE_MIN_SCALE,
                max_dim=self._config.IMAGE_MAX_DIM,
                mode=self._config.IMAGE_RESIZE_MODE)
            molded_image = self.mold_image(molded_image)
            # Build image_meta
            image_meta = self.compose_image_meta(
                0, image.shape, molded_image.shape, window, scale,
                np.zeros([self._config.NUM_CLASSES], dtype=np.int32))
            # Append
            molded_images.append(molded_image)
            windows.append(window)
            image_metas.append(image_meta)
        # Pack into arrays
        molded_images = np.stack(molded_images)
        image_metas = np.stack(image_metas)
        windows = np.stack(windows)
        return molded_images, image_metas, windows
```

That works well, allowing me to assemble batches of molded_images, image_metas, anchors and windows. Their shapes are now (8, 640, 640, 3), (8, 14), (8, 102300, 4), (8, 4), respectively (with batch size 8 and 2 classes).

Problem is that the tf model server gives me the following 7 errors:
```
2021-06-01 13:46:10.366846: W external/org_tensorflow/tensorflow/core/framework/op_kernel.cc:1651] OP_REQUIRES failed at strided_slice_op.cc:108 : Invalid argument: slice index 4 of dimension 0 out of bounds.
2021-06-01 13:46:10.366846: W external/org_tensorflow/tensorflow/core/framework/op_kernel.cc:1651] OP_REQUIRES failed at strided_slice_op.cc:108 : Invalid argument: slice index 7 of dimension 0 out of bounds.
2021-06-01 13:46:10.366868: W external/org_tensorflow/tensorflow/core/framework/op_kernel.cc:1651] OP_REQUIRES failed at strided_slice_op.cc:108 : Invalid argument: slice index 3 of dimension 0 out of bounds.
2021-06-01 13:46:10.366876: W external/org_tensorflow/tensorflow/core/framework/op_kernel.cc:1651] OP_REQUIRES failed at strided_slice_op.cc:108 : Invalid argument: slice index 6 of dimension 0 out of bounds.
2021-06-01 13:46:10.366880: W external/org_tensorflow/tensorflow/core/framework/op_kernel.cc:1651] OP_REQUIRES failed at strided_slice_op.cc:108 : Invalid argument: slice index 5 of dimension 0 out of bounds.
2021-06-01 13:46:10.366880: W external/org_tensorflow/tensorflow/core/framework/op_kernel.cc:1651] OP_REQUIRES failed at strided_slice_op.cc:108 : Invalid argument: slice index 2 of dimension 0 out of bounds.
2021-06-01 13:46:10.366880: W external/org_tensorflow/tensorflow/core/framework/op_kernel.cc:1651] OP_REQUIRES failed at strided_slice_op.cc:108 : Invalid argument: slice index 1 of dimension 0 out of bounds.
```

Obviously, the shape in the first dimension is 1, which is why only the first image in the batch is processed and the remaining 7 fail. I printed the tensor definitions in `export_saved_model` in `make_tf_serving.py` and get:
```
input_image tensor info name: "input_image:0"
dtype: DT_FLOAT
tensor_shape {
  dim {
    size: -1
  }
  dim {
    size: -1
  }
  dim {
    size: -1
  }
  dim {
    size: 3
  }
}

input_image_meta tensor info name: "input_image_meta:0"
dtype: DT_FLOAT
tensor_shape {
  dim {
    size: -1
  }
  dim {
    size: 14
  }
}

input_anchors tensor info name: "input_anchors:0"
dtype: DT_FLOAT
tensor_shape {
  dim {
    size: -1
  }
  dim {
    size: -1
  }
  dim {
    size: 4
  }
}

output_detection tensor info name: "mrcnn_detection/Reshape_1:0"
dtype: DT_INT64
tensor_shape {
  dim {
    size: -1
  }
}

output_mask tensor info name: "mrcnn_mask/Reshape_1:0"
dtype: DT_FLOAT
tensor_shape {
  dim {
    size: 8
  }
  dim {
    size: 200
  }
  dim {
    size: 28
  }
  dim {
    size: 28
  }
  dim {
    size: 2
  }
}
```

The shape of `mrcnn_detection/Reshape_1:0` seems to be off. When using your code with batch size 1, its shape is (1, 200, 6). The other shapes look similar to my scenario.

Do you have any idea how to resolve this issue? I would gladly appreciate any help on this. I guess, the model export has to be modified a bit.

BTW, I run the tensorflow server as follows:
```
docker run -it -p 8501:8501 --gpus all  -v "$(pwd)/exported_models:/models"  -e MODEL_NAME=mask_rcnn -t tensorflow/serving:1.15.0-gpu
```

**Comment by moganesyan** [comment]
It's been a while since I've worked on this but the first thing that comes to mind is the export config.

The MRCNN model has two parameters `IMAGES_PER_GPU` and `GPU_COUNT`. So the batch size is the product of the two.

You can override the default MRCNN parameters [here](https://github.com/moganesyan/tensorflow_model_deployment/blob/c34fe4cd840fc139a366508c8574af967e463c3e/tf_serving/make_tf_serving.py#L20)

**Comment by LukasBommes** [comment]
Thanks for your quick reply.
I have configured the batch size accordingly via
```
def get_model_config():
    """Returns the MS-COCO style inference config
    Ammend hyperparameters as necessary
    """
    config = InferenceConfig()
    config.GPU_COUNT = 1
    config.IMAGES_PER_GPU = 8
    config.BATCH_SIZE = 8
    return config
```

The problem seems to be related to the implementation of the `DetectionLayer` in line 782 of the Mask R-CNNN `model.py` file. I plotted all tensors and there are new tensors created with names Reshape_1:0, Reshape_2:0, ..., Reshape_8:0. While the first seven have shape (-1), the latter has the desired shape of (8, 200, 6). However, simply referring to this tensor in the `export_saved_model` didn't solve the problem. I get the same error message. Probably, because the other 7 tensors have shape (-1). I guess, it is somehow related to the `batch_slice` method applied in line 810 of `model.py` which creates copies of the Reshape_1:0 tensor.
